### DISCUSSION TITLE
Error when calling get_backend_symbol

### DISCUSSION THREAD

**Discussion by Zheweiqiu** [apache/incubator-mxnet#16042] [Q&A] [answered]
I am using the following code to use tensorrt to speed up inference but got some errors:
The mxnet model I am loading is retinaface downloaded from [Retinaface](https://github.com/deepinsight/insightface/tree/master/RetinaFace) under "RetinaFace Pretrained Models".
```
sym, arg_params, aux_params = mx.model.load_checkpoint(prefix, epoch)
trt_sym = sym.get_backend_symbol(‘TensorRT’)
mx.contrib.tensorrt.init_tensorrt_params(trt_sym, arg_params, aux_params)
mx.contrib.tensorrt.set_use_fp16(False)
self.model = trt_sym.simple_bind(ctx=self.ctx, data = (1,3,image_size[0], image_size[1]), grad_req=‘null’, force_rebind=True)
self.model.copy_params_from(arg_params, aux_params)
```

File “/home/qiuzhewei/RetinaFace/retinaface.py”, line 224, in init
trt_sym = sym.get_backend_symbol(‘TensorRT’)
File “/opt/conda/lib/python3.7/site-packages/mxnet/symbol/symbol.py”, line 2564, in get_backend_symbol
check_call(_LIB.MXGenBackendSubgraph(self.handle, c_str(backend), ctypes.byref(out)))
File “/opt/conda/lib/python3.7/site-packages/mxnet/base.py”, line 253, in check_call
raise MXNetError(py_str(LIB.MXGetLastError()))
mxnet.base.MXNetError: [09:39:00] src/c_api/…/operator/subgraph/subgraph_property.h:367: Check failed: it != prop_ptr_map.end(): SubgraphProperty TensorRT is not found in SubgraphPropertyRegistry
Stack trace:
[bt] (0) /opt/conda/lib/python3.7/site-packages/mxnet/libmxnet.so(+0x4a357b) [0x7f683cbe657b]
[bt] (1) /opt/conda/lib/python3.7/site-packages/mxnet/libmxnet.so(MXGenBackendSubgraph+0x1ab) [0x7f683ed5a5db]
[bt] (2) /opt/conda/lib/python3.7/lib-dynload/…/…/libffi.so.6(ffi_call_unix64+0x4c) [0x7f6876a4aec0]
[bt] (3) /opt/conda/lib/python3.7/lib-dynload/…/…/libffi.so.6(ffi_call+0x22d) [0x7f6876a4a87d]
[bt] (4) /opt/conda/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so(_ctypes_callproc+0x2ce) [0x7f6876ec9f7e]
[bt] (5) /opt/conda/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so(+0x139b4) [0x7f6876eca9b4]
[bt] (6) python(_PyObject_FastCallKeywords+0x49b) [0x55b2e1d13d2b]
[bt] (7) python(_PyEval_EvalFrameDefault+0x537e) [0x55b2e1d6f7ae]
[bt] (8) python(_PyFunction_FastCallKeywords+0xfb) [0x55b2e1d1279b]

The code is basically from /tests/python/tensorrt/test_resnet18.py
I have no idea how to solve it and google doesn't provide any solution.
Any help will be appreciate!

**Comment by ZhennanQin** [comment]
TensorRT integration isn't built by default. To enable this backend, you need to build from source with `USE_TENSORRT=1`

**Comment by leleamol** [comment]
@Zheweiqiu please let us know if building with TENSORRT solves the issue.

@mxnet-label-bot add [Question, Build, Pending Requester Info]

**Comment by Zheweiqiu** [comment]
Hi @ZhennanQin,
When you say "build from source", do you mean set parameter `USE_TENSORRT=1` when building the model or do I need to reinstall mxnet library with tensorrt support? I guess it is the former one since I can successfully call `mxnet.contrib.tensorrt `.

**Comment by ZhennanQin** [comment]
I mean you need "to reinstall mxnet library with tensorrt support".  But if you can "successfully call mxnet.contrib.tensorrt", then I assume you're using the build that enable TensorRT backend. Then please try to add `os.environ['MXNET_USE_TENSORRT'] = '1'`  in your scripts. Please read this documentation for full details:
https://github.com/apache/incubator-mxnet/blob/master/docs/tutorials/tensorrt/inference_with_trt.md

**Comment by Zheweiqiu** [comment]
> I mean you need "to reinstall mxnet library with tensorrt support". But if you can "successfully call mxnet.contrib.tensorrt", then I assume you're using the build that enable TensorRT backend. Then please try to add `os.environ['MXNET_USE_TENSORRT'] = '1'` in your scripts. Please read this documentation for full details:
> https://github.com/apache/incubator-mxnet/blob/master/docs/tutorials/tensorrt/inference_with_trt.md

Below is what I have tried
My initial code: 
```
sym, arg_params, aux_params = mx.model.load_checkpoint(prefix, epoch)
if not use_trt:
        self.model = mx.mod.Module(symbol=sym, context=self.ctx, label_names = None)
        self.model.bind(data_shapes=[('data', (1, 3, image_size[0], image_size[1]))], for_training=False)
        self.model.set_params(arg_params, aux_params)
else:
        print('------------ Using tensorrt for face detection --------------')
        trt_sym = sym.get_backend_symbol('TensorRT')
        mx.contrib.tensorrt.init_tensorrt_params(trt_sym, arg_params, aux_params)
        mx.contrib.tensorrt.set_use_fp16(False)
        self.model = trt_sym.simple_bind(ctx=self.ctx, data = (1,3,image_size[0], image_size[1]), grad_req='null', force_rebind=True)
        self.model.copy_params_from(arg_params, aux_params)

```
I added one line according to your answer
```
sym, arg_params, aux_params = mx.model.load_checkpoint(prefix, epoch)
if not use_trt:
        self.model = mx.mod.Module(symbol=sym, context=self.ctx, label_names = None)
        self.model.bind(data_shapes=[('data', (1, 3, image_size[0], image_size[1]))], for_training=False)
        self.model.set_params(arg_params, aux_params)
else:
        print('------------ Using tensorrt for face detection --------------')
        os.environ['MXNET_USE_TENSORRT'] = '1'
        trt_sym = sym.get_backend_symbol('TensorRT')
        mx.contrib.tensorrt.init_tensorrt_params(trt_sym, arg_params, aux_params)
        mx.contrib.tensorrt.set_use_fp16(False)
        self.model = trt_sym.simple_bind(ctx=self.ctx, data = (1,3,image_size[0], image_size[1]), grad_req='null', force_rebind=True)
        self.model.copy_params_from(arg_params, aux_params)

```
But I still got the error:
mxnet.base.MXNetError: [06:56:32] src/c_api/../operator/subgraph/subgraph_property.h:367: Check failed: it != prop_ptr_map_.end(): SubgraphProperty TensorRT is not found in SubgraphPropertyRegistry

Then I tried alternative in that document and changed my code to:

```
sym, arg_params, aux_params = mx.model.load_checkpoint(prefix, epoch)
if not use_trt:
        self.model = mx.mod.Module(symbol=sym, context=self.ctx, label_names = None)
        self.model.bind(data_shapes=[('data', (1, 3, image_size[0], image_size[1]))], for_training=False)
        self.model.set_params(arg_params, aux_params)
else:
        print('------------ Using tensorrt for face detection --------------')
        os.environ['MXNET_USE_TENSORRT'] = '1'
        arg_params.update(aux_params)
        all_params = dict([(k, v.as_in_context(self.ctx)) for k,v in arg_params.items()])
        self.model = mx.contrib.tensorrt.tensorrt_bind(sym, ctx=self.ctx, all_params=all_params, data=batch_shape, grad_req='null', force_rebind=True)

```

But still got an errro:
AttributeError: module 'mxnet.contrib.tensorrt' has no attribute 'tensorrt_bind'

Below is my mxnet information:
Name: mxnet-cu100mkl
Version: 1.5.0
Summary: MXNet is an ultra-scalable deep learning framework. This version uses CUDA-10.0 and MKLDNN.
Home-page: https://github.com/apache/incubator-mxnet
Author: UNKNOWN
Author-email: UNKNOWN
License: Apache 2.0
Location: /opt/conda/lib/python3.7/site-packages
Requires: numpy, requests, graphviz
Required-by:

Is that because I am using wrong version of mxnet?

Appreciate for any help!

**Comment by ptrendx** [comment]
@KellenSunderland @Caenorst

**Comment by Zheweiqiu** [comment]
> @Zheweiqiu please let us know if building with TENSORRT solves the issue.
> 
> @mxnet-label-bot add [Question, Build, Pending Requester Info]

Still facing the same problem after trying approaches listed above, can you take a look at it?
Thanks in advance!

**Comment by Caenorst** [comment]
@Zheweiqiu Can you show the whole process / script that you did for building MXNet ?

**Comment by Zheweiqiu** [comment]
> @Zheweiqiu Can you show the whole process / script that you did for building MXNet ?

I am using retinaface mxnet model and the code is basically from [github](https://github.com/deepinsight/insightface/blob/master/RetinaFace/retinaface.py). I want to speed up the reference using tensorrt, so I modify it a little bit.
The whole scrip is too long so I only attach modification I make. Below is original code from line 113 to 138.

```
    sym, arg_params, aux_params = mx.model.load_checkpoint(prefix, epoch)
    if self.ctx_id>=0:
      self.ctx = mx.gpu(self.ctx_id)
      self.nms = gpu_nms_wrapper(self.nms_threshold, self.ctx_id)
    else:
      self.ctx = mx.cpu()
      self.nms = cpu_nms_wrapper(self.nms_threshold)
    self.pixel_means = np.array(pixel_means, dtype=np.float32)
    self.pixel_stds = np.array(pixel_stds, dtype=np.float32)
    self.pixel_scale = float(pixel_scale)
    print('means', self.pixel_means)
    self.use_landmarks = False
    if len(sym)//len(self._feat_stride_fpn)==3:
      self.use_landmarks = True
    print('use_landmarks', self.use_landmarks)

    if self.debug:
      c = len(sym)//len(self._feat_stride_fpn)
      sym = sym[(c*0):]
      self._feat_stride_fpn = [32,16,8]
    print('sym size:', len(sym))

    image_size = (640, 640)
    self.model = mx.mod.Module(symbol=sym, context=self.ctx, label_names = None)
    self.model.bind(data_shapes=[('data', (1, 3, image_size[0], image_size[1]))], for_training=False)
    self.model.set_params(arg_params, aux_params)
```

I change it to 

```
        if use_trt:
            os.environ['MXNET_USE_TENSORRT'] = '1' 
        sym, arg_params, aux_params = mx.model.load_checkpoint(prefix, epoch)
        if self.ctx_id>=0:
            self.ctx = mx.gpu(self.ctx_id)
            self.nms = gpu_nms_wrapper(self.nms_threshold, self.ctx_id)
        else:
            self.ctx = mx.cpu()
            self.nms = cpu_nms_wrapper(self.nms_threshold)
        self.pixel_means = np.array(pixel_means, dtype=np.float32)
        self.pixel_stds = np.array(pixel_stds, dtype=np.float32)
        self.pixel_scale = float(pixel_scale)
        #print('means', self.pixel_means)
        self.use_landmarks = False
        if len(sym)//len(self._feat_stride_fpn)==3:
            self.use_landmarks = True
        #print('use_landmarks', self.use_landmarks)

        if self.debug:
            c = len(sym)//len(self._feat_stride_fpn)
            sym = sym[(c*0):]
            self._feat_stride_fpn = [32,16,8]
        #print('sym size:', len(sym))

        image_size = (640, 640)
        if not use_trt:
            self.model = mx.mod.Module(symbol=sym, context=self.ctx, label_names = None)
            self.model.bind(data_shapes=[('data', (1, 3, image_size[0], image_size[1]))], for_training=False)
            self.model.set_params(arg_params, aux_params)
        else:
            print('------------ Using tensorrt for face detection --------------')
            #''' 
            trt_sym = sym.get_backend_symbol('TensorRT')
            mx.contrib.tensorrt.init_tensorrt_params(trt_sym, arg_params, aux_params)
            mx.contrib.tensorrt.set_use_fp16(False)
            self.model = trt_sym.simple_bind(ctx=self.ctx, data = (1,3,image_size[0], image_size[1]), grad_req='null', force_rebind=True)
            self.model.copy_params_from(arg_params, aux_params)
            '''
            arg_params.update(aux_params)
            all_params = dict([(k, v.as_in_context(self.ctx)) for k,v in arg_params.items()])
            self.model = mx.contrib.tensorrt.tensorrt_bind(sym, ctx=self.ctx, all_params=all_params, data=batch_shape, grad_req='null', force_rebind=True)
            '''
```
When I set use_trt=True, I get the following error:

[32, 16, 8] {'32': {'SCALES': (32, 16), 'BASE_SIZE': 16, 'RATIOS': (1.0,), 'ALLOWED_BORDER': 9999}, '16': {'SCALES': (8, 4), 'BASE_SIZE': 16, 'RATIOS': (1.0,), 'ALLOWED_BORDER': 9999}, '8': {'SCALES': (2, 1), 'BASE_SIZE': 16, 'RATIOS': (1.0,), 'ALLOWED_BORDER': 9999}}
[07:01:19] src/nnvm/legacy_json_util.cc:209: Loading symbol saved by previous version v1.3.0. Attempting to upgrade...
[07:01:19] src/nnvm/legacy_json_util.cc:217: Symbol successfully upgraded!
------------ Using tensorrt for face detection --------------
Traceback (most recent call last):
  File "test_server_flask_celebrity.py", line 32, in <module>
    celebrity = cm_model.CelebModel(args)
  File "/home/qiuzhewei/celeb_test/src/test_celeb_model.py", line 71, in __init__
    detector = RetinaFace(args.detectmodel, 0, args.gpu, 'net3')
  File "/home/qiuzhewei/celeb_test/RetinaFace/retinaface.py", line 227, in __init__
    trt_sym = sym.get_backend_symbol('TensorRT')
  File "/opt/conda/lib/python3.7/site-packages/mxnet/symbol/symbol.py", line 2564, in get_backend_symbol
    check_call(_LIB.MXGenBackendSubgraph(self.handle, c_str(backend), ctypes.byref(out)))
  File "/opt/conda/lib/python3.7/site-packages/mxnet/base.py", line 253, in check_call
    raise MXNetError(py_str(_LIB.MXGetLastError()))
mxnet.base.MXNetError: [07:01:19] src/c_api/../operator/subgraph/subgraph_property.h:367: Check failed: it != prop_ptr_map_.end(): SubgraphProperty TensorRT is not found in SubgraphPropertyRegistry
Stack trace:
  [bt] (0) /opt/conda/lib/python3.7/site-packages/mxnet/libmxnet.so(+0x4a357b) [0x7f9b1a05d57b]
  [bt] (1) /opt/conda/lib/python3.7/site-packages/mxnet/libmxnet.so(MXGenBackendSubgraph+0x1ab) [0x7f9b1c1d15db]
  [bt] (2) /opt/conda/lib/python3.7/lib-dynload/../../libffi.so.6(ffi_call_unix64+0x4c) [0x7f9b53e81ec0]
  [bt] (3) /opt/conda/lib/python3.7/lib-dynload/../../libffi.so.6(ffi_call+0x22d) [0x7f9b53e8187d]
  [bt] (4) /opt/conda/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so(_ctypes_callproc+0x2ce) [0x7f9b54300f7e]
  [bt] (5) /opt/conda/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so(+0x139b4) [0x7f9b543019b4]
  [bt] (6) python(_PyObject_FastCallKeywords+0x49b) [0x55b7bc065d2b]
  [bt] (7) python(_PyEval_EvalFrameDefault+0x537e) [0x55b7bc0c17ae]
  [bt] (8) python(_PyFunction_FastCallKeywords+0xfb) [0x55b7bc06479b]

And I have no idea how to solve it. Any help will be appreciate!

**Comment by Caenorst** [comment]
@Zheweiqiu Sorry, I asked you about the process for BUILDING MXNet.

**Comment by Zheweiqiu** [comment]
> @Zheweiqiu Sorry, I asked you about the process for BUILDING MXNet.

You mean the process of building model? I am using the pretrained model from [github-retinaface](https://github.com/deepinsight/insightface/tree/master/RetinaFace) under "RetinaFace Pretrained Models" section

**Comment by Caenorst** [comment]
We suggested your rebuild MXNet from Source with TensorRT and you said you did it. Did you ?

**Comment by grapeot** [comment, marked as answer]
@Caenorst I can see a misunderstanding here that you meant to build MXNet with TensorRT support but @Zheweiqiu thought it meant to build the model. I hit this problem with exactly the same error messages using the default python package, and am willing to dig deeper by rebuilding MXNet with TensorRT support. ~~However, I don't see instructions in the [doc](https://github.com/apache/incubator-mxnet/blob/1.5.1/docs/install/build_from_source.md), or the [make options](https://github.com/apache/incubator-mxnet/blob/master/make/config.mk). Wondering do you mind pointing us to the proper doc? Thanks for your time!~~

Update: found the switch USE_TENSORRT in [Cmake file](https://github.com/apache/incubator-mxnet/blob/1.5.1/CMakeLists.txt).
Update: found this auto build [script](https://github.com/apache/incubator-mxnet/blob/master/ci/docker/runtime_functions.sh#L679).
Update: was able to successfully build MXNet with TensorRT support by following the aforementioned script (used CUDA 10.1, CuDNN 7.6, TensorRT 6.0). But the code got stuck for half an hour with 100% CPU on one core, before it threw a ton of error messages on some nodes have a cycle. Anyway the original error message disappeared. But I was trying on a different model. @Zheweiqiu you may want to try this approach. It may work on your model.
